File: app/src/main/java/com/mainstreetcode/teammate/adapters/TeamEditAdapter.java
package com.mainstreetcode.teammate.adapters;

import android.view.ViewGroup;

import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.adapters.viewholders.input.InputViewHolder;
import com.mainstreetcode.teammate.adapters.viewholders.input.SpinnerTextInputStyle;
import com.mainstreetcode.teammate.adapters.viewholders.input.TextInputStyle;
import com.mainstreetcode.teammate.baseclasses.BaseAdapter;
import com.mainstreetcode.teammate.baseclasses.BaseViewHolder;
import com.mainstreetcode.teammate.fragments.headless.ImageWorkerFragment;
import com.mainstreetcode.teammate.model.Config;
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable;
import com.mainstreetcode.teammate.model.Item;
import com.mainstreetcode.teammate.model.Team;
import com.mainstreetcode.teammate.model.enums.Sport;
import com.mainstreetcode.teammate.util.ViewHolderUtil;

import java.util.List;

import androidx.annotation.NonNull;

import static com.mainstreetcode.teammate.model.Item.ZIP;
import static com.mainstreetcode.teammate.util.ViewHolderUtil.ITEM;

/**
 * Adapter for {@link Team}
 */

public class TeamEditAdapter extends BaseAdapter<InputViewHolder, TeamEditAdapter.TeamEditAdapterListener> {

    private final List<Differentiable> items;
    private final TextInputStyle.InputChooser chooser;

    public TeamEditAdapter(List<Differentiable> items, TeamEditAdapter.TeamEditAdapterListener listener) {
        super(listener);
        // setHasStableIds(true); DO NOT PUT THIS BACK
        this.items = items;
        chooser = new Chooser(adapterListener);
    }

    @NonNull
    @Override
    public InputViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return new InputViewHolder(getItemView(R.layout.viewholder_simple_input, viewGroup));
    }

    @SuppressWarnings("unchecked")
    @Override protected <S extends AdapterListener> S updateListener(BaseViewHolder<S> viewHolder) {
        return (S) adapterListener;
    }

    @Override
    public void onBindViewHolder(@NonNull InputViewHolder viewHolder, int i) {
        super.onBindViewHolder(viewHolder, i);
        Differentiable item = items.get(i);
        if (item instanceof Item) viewHolder.bind(chooser.get((Item) item));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public int getItemViewType(int position) {
        return ITEM;
    }

    public interface TeamEditAdapterListener extends ImageWorkerFragment.ImagePickerListener {

        void onAddressClicked();

        boolean canEditFields();
    }

    private static class Chooser extends TextInputStyle.InputChooser {

        private TeamEditAdapterListener adapterListener;

        Chooser(TeamEditAdapterListener adapterListener) {
            this.adapterListener = adapterListener;
        }

        @Override public int iconGetter(Item item) {
            return item.getStringRes() == R.string.team_name && adapterListener.canEditFields()
                    ? R.drawable.ic_picture_white_24dp : 0;
        }

        @Override public boolean enabler(Item item) {
            switch (item.getItemType()) {
                default:
                case Item.ABOUT:
                    return Item.FALSE.apply(item);
                case Item.ZIP:
                case Item.CITY:
                case Item.INFO:
                case Item.STATE:
                case Item.INPUT:
                case Item.SPORT:
                case Item.NUMBER:
                case Item.DESCRIPTION:
                    return adapterListener.canEditFields();
            }
        }

        @Override public CharSequence textChecker(Item item) {
            switch (item.getItemType()) {
                default:
                case Item.CITY:
                case Item.STATE:
                case Item.INPUT:
                case Item.NUMBER:
                    return Item.NON_EMPTY.apply(item);
                case Item.INFO:
                    return ViewHolderUtil.allowsSpecialCharacters.apply(item.getValue());
                case Item.ZIP:
                case Item.DESCRIPTION:
                    return Item.ALL_INPUT_VALID.apply(item);
            }
        }

        @Override public TextInputStyle apply(Item item) {
            int itemType = item.getItemType();
            switch (itemType) {
                default:
                case Item.ZIP:
                case Item.CITY:
                case Item.INFO:
                case Item.STATE:
                case Item.ABOUT:
                case Item.INPUT:
                case Item.NUMBER:
                case Item.DESCRIPTION:
                    return new TextInputStyle(
                            itemType == Item.CITY || itemType == Item.STATE || itemType == ZIP
                                    ? adapterListener::onAddressClicked
                                    : Item.NO_CLICK,
                            itemType == Item.INPUT
                                    ? adapterListener::onImageClick
                                    : adapterListener::onAddressClicked,
                            this::enabler,
                            this::textChecker,
                            this::iconGetter);
                case Item.SPORT:
                    return new SpinnerTextInputStyle<>(
                            R.string.choose_sport,
                            Config.getSports(),
                            Sport::getName,
                            Sport::getCode,
                            this::enabler);
            }
        }
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/adapters/TeamEditAdapter.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.adapters

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.adapters.viewholders.input.InputViewHolder
import com.mainstreetcode.teammate.adapters.viewholders.input.SpinnerTextInputStyle
import com.mainstreetcode.teammate.adapters.viewholders.input.TextInputStyle
import com.mainstreetcode.teammate.adapters.viewholders.input.or
import com.mainstreetcode.teammate.fragments.headless.ImageWorkerFragment
import com.mainstreetcode.teammate.model.Config
import com.mainstreetcode.teammate.model.Item
import com.mainstreetcode.teammate.model.enums.Sport
import com.mainstreetcode.teammate.model.neverEnabled
import com.mainstreetcode.teammate.model.noBlankFields
import com.mainstreetcode.teammate.model.noInputValidation
import com.mainstreetcode.teammate.model.noSpecialCharacters
import com.tunjid.androidx.recyclerview.adapterOf
import com.tunjid.androidx.recyclerview.diff.Differentiable
import com.tunjid.androidx.view.util.inflate

interface TeamEditAdapterListener : ImageWorkerFragment.ImagePickerListener {

    fun onAddressClicked()

    fun canEditFields(): Boolean
}

fun teamEditAdapter(
        modelSource: () -> List<Differentiable>,
        delegate: TeamEditAdapterListener
): RecyclerView.Adapter<InputViewHolder> {
    val chooser = TeamEditChooser(delegate)
    return adapterOf(
            itemsSource = modelSource,
            viewHolderCreator = { viewGroup: ViewGroup, _: Int ->
                InputViewHolder(viewGroup.inflate(R.layout.viewholder_simple_input))
            },
            viewHolderBinder = { holder, item, _ ->
                if (item is Item) holder.bind(chooser[item])
            },
            onViewHolderRecycled = InputViewHolder::clear,
            onViewHolderDetached = InputViewHolder::onDetached,
            onViewHolderRecycleFailed = { it.clear(); false }
    )
}

private class TeamEditChooser internal constructor(
        private val delegate: TeamEditAdapterListener
) : TextInputStyle.InputChooser() {

    override fun iconGetter(item: Item): Int = when {
        item.stringRes == R.string.team_name && delegate.canEditFields() -> R.drawable.ic_picture_white_24dp
        else -> 0
    }

    override fun enabler(item: Item): Boolean = when (item.itemType) {
        Item.ABOUT -> item.neverEnabled
        Item.ZIP,
        Item.CITY,
        Item.INFO,
        Item.STATE,
        Item.INPUT,
        Item.SPORT,
        Item.NUMBER,
        Item.DESCRIPTION -> delegate.canEditFields()
        else -> item.neverEnabled
    }

    override fun textChecker(item: Item): CharSequence? = when (item.itemType) {
        Item.CITY,
        Item.STATE,
        Item.INPUT,
        Item.NUMBER -> item.noBlankFields
        Item.INFO -> item.noSpecialCharacters
        Item.ZIP,
        Item.DESCRIPTION -> item.noInputValidation
        else -> item.noBlankFields
    }

    override fun invoke(item: Item): TextInputStyle = when (val itemType = item.itemType) {
        Item.ZIP,
        Item.CITY,
        Item.INFO,
        Item.STATE,
        Item.ABOUT,
        Item.INPUT,
        Item.NUMBER,
        Item.DESCRIPTION -> TextInputStyle(
                or((itemType == Item.CITY || itemType == Item.STATE || itemType == Item.ZIP),
                        delegate::onAddressClicked,
                        Item.noClicks),
                or(itemType == Item.INPUT,
                        delegate::onImageClick,
                        delegate::onAddressClicked),
                this::enabler,
                this::textChecker,
                this::iconGetter)
        Item.SPORT -> SpinnerTextInputStyle(
                R.string.choose_sport,
                Config.getSports(),
                Sport::name,
                Sport::code,
                this::enabler)
        else -> TextInputStyle(
                or((itemType == Item.CITY || itemType == Item.STATE || itemType == Item.ZIP),
                        delegate::onAddressClicked,
                        Item.noClicks),
                or(itemType == Item.INPUT,
                        delegate::onImageClick,
                        delegate::onAddressClicked),
                this::enabler,
                this::textChecker,
                this::iconGetter)
    }
}